### ISSUE TITLE
How does ACM determine the device group?

### ISSUE THREAD

**Issue by FKuypers** [Ylianst/MeshCentral#952]
I am trying to activate a device in ACM, but it basically fails with the server message:

> Server error: Invalid device group

This device has once been provisioned on a different domain on this server and a different group, but I have deleted the system from this domain and group.

Nevertheless, when I use the original group for ACM, the activation proceeds (only in CCM, since the DNS Suffix is wrong)

I thought there would be some resemblance in the database, but I cannot really see the UUID of the device referenced somehow.

Why and where does the server check the device being linked to a specific group ? The activation string should define the group, shouldn't it?

How can I reactivate a device that has been part of a different group/domain once but since then has been deleted?

Thanks,

Frank

**Comment by Ylianst** [comment]
Looking into this now...

**Comment by Ylianst** [comment]
Just took a look at this. I presume you are activating using `meshcmd` with a command line like this:

`meshcmd amtacm --url wss://devbox.mesh.meshcentral.com/amtactivate?id=sjN2rO8... --serverhttpshash ABD3...`

The value that follows id= is the device group identifier, but the URL itself will indicate the domain. Does you second domain have a DNS name associated with it? For example I have a second domain called "devtest" and so, my URL is:

`meshcmd amtccm --url wss://devbox.mesh.meshcentral.com/devtest/amtactivate?id=XJE... --serverhttpshash ABD39...`

Note the "devtest" in the URL. If I set a DNS name for that domain, I will get a URL like this:

`meshcmd amtccm --url wss://devtest.mesh.meshcentral.com/amtactivate?id=XJE... --serverhttpshash ABD39...`

Let me know what situation you are in. MeshCentral can't seem to find the device group based on the URL and ID.

You can also type:

`node node_modules/meshcentral --showmeshes`

This will dump all of the device groups in JSON so you can look at it. Lastly, I am changing the code so on the next update of MeshCentral, the error message will be "Invalid device group: [mesh/domain/id]" so it will be easy to debug what is going on.

**Comment by FKuypers** [comment]
Ok,

I got 2 Meshes in this domain:

`  {
    type: 'mesh',
    _id: 'mesh/demo/NDEATlAXhc40e3zgi8rcCTQ@3ZD@Jy0tsn@8izghahM4MC$dY8DMH4oB6VS8VsyR',
    name: 'AMT Only',
    mtype: '1',
    desc: 'This group contains clients only managed with AMT',
    domain: 'demo',
    links: { 'user/demo/frank': [Object] },
    amt: { type: 3, password: 'xxxxxxxx', cirasetup: 2 }
  },
  {
    type: 'mesh',
    _id: 'mesh/demo/R2iymM3pUMfWHCqE5bdVwhcsGx$ZCvbs4sbtJyvoTyKiM000wtdAM3k9oquwou8H',
    name: 'Agent+AMT',
    mtype: '2',
    desc: 'Manages clients using the Agent and AMT for OOBM',
    domain: 'demo',
    links: { 'user/demo/frank': [Object] },
    flags: 2,
    amt: { type: 3, password: 'xxxxxxxxx', cirasetup: 2 }
  }
`
It does use a subdomain as you can see. I grab the acticvation string from the "AMT-Only" mesh, because the agent itself does not activate AMT - maybe because of the same reason.

`meshcmd amtacm --url wss://demo.imu-lab.de/amtactivate?id=NDEATlAXhc40e3zgi8rcCTQ@3ZD@Jy0tsn@8izghahM4MC$dY8DMH4oB6VS8VsyR --serverhttpshash D89BB87DCB5FD26CD4832534C318DECC0FF1208D1640FFA62206BDCE63999A8945B3940F126C172DF23A66A8BAC4254C`

I replace the ID with the other mesh-id:

`meshcmd amtacm --url wss://demo.imu-lab.de/amtactivate?id=**R2iymM3pUMfWHCqE5bdVwhcsGx$ZCvbs4sbtJyvoTyKiM000wtdAM3k9oquwou8H** --serverhttpshash D89BB87DCB5FD26CD4832534C318DECC0FF1208D1640FFA62206BDCE63999A8945B3940F126C172DF23A66A8BAC4254C`

Then I get the error message as stated above.

I will wait for the next version, the affected system is now in the office - and I am not anymore ;-) 

Stay tuned....

Frank

**Comment by Ylianst** [comment]
I published MeshCentral v0.4.9-j with the improved error message. I am wondering if the "$" symbol in the identifier if causing the problem. When you update to the new version, the "Invalid device group" error message may confirm this.

**Comment by FKuypers** [comment]
Well,

it definitely caused a problem in Powershell - hence I typed CMD to get around that problem.

Both Meshes have a "$" in them - when I simply use the AMT-Only Mesh, it does work... So I do not think it is caused by the $ sign...

Frank

**Comment by Ylianst** [comment]
I just setup my developer machine in the same situation you have and got the same error! Hold on, working on it.

**Comment by Ylianst** [comment]
ARG! I see the problem... I will get you a new MeshCentral shortly.

**Comment by Ylianst** [comment]
Published MeshCentral v0.4.9-k with a fix for this. Probably also fixes other issued related to having a second domain that is DNS based. Update your server and let me know if it fixes it.

**Comment by FKuypers** [comment]
Thanks for the super fast turnaround - I will let you know - but it might take a day or two before I have access to the system again. 

Frank

**Comment by FKuypers** [comment]
I just unprovisioned another maschine I had handy and tried again - only difference, this time it's on the main domain and not a sub domain:

`Server error: Invalid device group type:mesh//DXC$lUn8CSFGrrFZ4Qz9RpIADaPNhNJdp9yhQ1@r4WH4mnVCPEYOEMqcAxfp$l7V`

Frank

**Comment by Ylianst** [comment]
Your getting "Invalid group type". Most likely your trying to activate using a device group that is with agent, not the "Intel AMT Only / Agent-less" group type.

**Comment by FKuypers** [comment]
Ok, 
maybe I did not grasp the concept then...

I have a group that uses an agent, but also I want AMT to be used to I set ACM + CIRA as well...

After the install of the agent, how is AMT activated then if not by MeshCmd?

Frank

**Comment by Ylianst** [comment]
Normally, when you have a group setup to use agents, the agent should do all of the ACM activation work for you. The agent should try to see if there is an opportunity for ACM activation and do it when possible. Only when setting up agent-less do you need to run MeshCMD.

When setting up with agent, the server will also arrange to match the CIRA connection with the agent, so they both show up until the same device on the web site.

This said, if you find issues, let me know. I am going on vacation for a week starting tomorrow, but I do want to fix any problem you encounter in this area.

**Comment by FKuypers** [comment]
OK,

I believe one of my problems is, that I am using Wifi clients. I have not seen any activation activity with them, I did have a LAN based computer, that activated after reboot on its own - but still the password was not set for the credentials for AMT to connect.

I tried the AMT to add the client (Wifi based) to add to the AMT only group, and that worked using the MeshCMD command line - but even though the Simple ACM profile called for CIRA to be enabled, this was never written into the ME.

So I loaded the CIRA Script using MeshCommander, set my Wifi Credentials using the AMT Webinterface and finally set the admin password which I grabbed from the AMTActivation.log an added it to my the settings in the group, where the client was sitting for Agent/AMT management and it all worked.

That leaves me with a few questions:

- Can I trigger the agent to start the ACM configuration and see results? The agent console does not seem to do the trick for me? Is there a log file for the agent somewhere?
- CIRA setting to not seem to get written during activation - is this done in a later stage or do I always need to use a Meshcommander script?
- Is there a way to write WIFI credentials/Enable Wifi using a script?

 Thanks,

Frank

**Comment by FKuypers** [comment]
Sorry for comming back this late...

I tried around a bit more. I do use LAN now.

I created 3 device groups:

- AMT ACM only
- Agent + ACM
- Agent + CCM

I can activate using MeshCmd on the client getting the system into the "AMT ACM only" group. But because I wanted a random password, there is no way to retrieve the admin password from the web UI to install the CIRA script to actually have the system connect to MeshCentral. The password is stored in the log file of the server, but not in the database for the client, so once I do set the CIRA settings using the MeshCommander script, the client still can't be managed due to the missing admin password.

If I unconfigure the system and install an agent for one of the other groups above, I do not see any provisioning happening. 

There is an uncaughtException1: Error: => EventEmitter.emit() in the agent console of the client - maybe this is related?

Can the agent be triggered to perfrom AMT activation - either in CCM or ACM mode?

Thanks,
Frank
